packages/next/src/build/webpack/plugins/next-types-plugin.ts

```diff
@@ -118,17 +127,65 @@ async function collectNamedSlots(layoutPath: string) {
   return slots
 }
 
-const nodeRouteTypes: string[] = []
-const edgeRouteTypes: string[] = []
+const nodeRoutes: string[] = []
+const edgeRoutes: string[] = []
+
+export const devPageFiles = new Set<string>()
+
+function routeToRouteType(route: string) {
+  if (isDynamicRoute(route)) {
+    route = route
+      .split('/')
+      .map((part) => {
+        if (part.startsWith('[') && part.endsWith(']')) {
+          if (part.startsWith('[...')) {
+            // /[...slug]
+            return `\${CatchAllSlug<T>}`
+          } else if (part.startsWith('[[...') && part.endsWith(']]')) {
+            // /[[...slug]]
+            return `\${OptionalCatchAllSlug<T>}`
+          }
+          // /[slug]
+          return `\${SafeSlug<T>}`
+        }
+        return part
+      })
+      .join('/')
+  }
+
+  return `\`${route}\``
+}
 
-export const pageFiles = new Set<string>()
+function createRouteDefinitions(
+  originalNextModule: string,
+  rewrites: Rewrites,
+  redirects: Redirect[]
+) {
+  const extraRoutes: string[] = []
+  function addExtraRoute(route: string) {
+    // TODO: For now, we only handle absolute rewrite sources strictly, and ignore
+    // these with parameters or regexps.
+    if (!/\/:/.test(route)) {
```

Comment:
There can also be un-named matches e.g. `/hello/(.*)` potentially we can use the `Tokens` similar to how we check segments in `load-custom-routes` 
x-ref: https://github.com/vercel/next.js/blob/624232d064a2b7f0389746ca7c6b6944433bf566/packages/next/src/lib/load-custom-routes.ts#L348